### DISCUSSION TITLE
How to bypass `No keyring file or directory provided` error?

### DISCUSSION THREAD

**Discussion by tigerinus** [rauc/rauc#1207] [Q&A] [answered]
Hello,

I read from https://rauc.readthedocs.io/en/latest/using.html#creating-bundles that

> While the --cert and --key argument are mandatory for signing and must provide the certificate and private key that should be used for creating the signature, the --keyring argument is **optional** and (if given) will be used for verifying the trust chain validity of the signature after creation. Note that this is very useful to prevent signing with obsolete certificates, etc.

Therefore in my building step I don't have this `--keyring` argument specified:

```bash
function create_ota_update() {
    local ota_file="$(casaos_image_name raucb)"
    local ota_compatible="$(casaos_rauc_compatible)"
    local ota_version="$(casaos_version)"
    local rauc_folder="${BINARIES_DIR}/rauc"
    local boot="${BINARIES_DIR}/boot.vfat"
    local kernel="${BINARIES_DIR}/kernel.img"
    local rootfs="${BINARIES_DIR}/rootfs.squashfs"
    local spl="${BINARIES_DIR}/spl.img"

    if [ "${DEPLOYMENT}" == "production" ]; then
        local key="/build/ota_key_prod.pem"
        local cert="/build/ota_cert_prod.pem"
    else
        local key="/build/ota_key_dev.pem"
        local cert="/build/ota_cert_dev.pem"
    fi

    # Skip if no dev key is arround
    if [ ! -f "${key}" ]; then
        echo "Skip creating OTA update because of missing key ${key}"
        return 0
    fi

    rm -rf "${rauc_folder}" "${ota_file}"
    mkdir -p "${rauc_folder}"

    cp -f "${kernel}" "${rauc_folder}/kernel.img"
    cp -f "${boot}" "${rauc_folder}/boot.vfat"
    cp -f "${rootfs}" "${rauc_folder}/rootfs.img"
    cp -f "${BR2_EXTERNAL_CASAOS_PATH}/ota/rauc-hook" "${rauc_folder}/hook"

    # SPL
    if [ "${BOOT_SPL}" == "true" ]; then
        cp -f "${spl}" "${rauc_folder}/spl.img"
    fi

    export BOOTLOADER BOOT_SPL
    export ota_compatible ota_version
    (
        "${HOST_DIR}/bin/tempio" \
            -template "${BR2_EXTERNAL_CASAOS_PATH}/ota/manifest.raucm.gtpl"
    ) > "${rauc_folder}/manifest.raucm"

    rauc bundle -d --cert="${cert}" --key="${key}" "${rauc_folder}" "${ota_file}"
}
```

However, when running `rauc install ...`, it still complains about `No keyring file or directory provided` 

```
installing
  0% Installing
  0% Determining slot states
 20% Determining slot states done.
 20% Checking bundle
 40% Checking bundle failed.
100% Installing failed.
LastError: No keyring file or directory provided
```

Please help! Thanks.

**Comment by ejoerns** [comment]
The keyring is optional only for the bundle *creation* as post-creation verification is not mandatory.

However, when installing, bundle verification cannot be optional (otherwise it would not make sense to sign an update).

So what you have to do is place your keyring file in the rootfs and configure this location in your [system.conf](https://rauc.readthedocs.io/en/latest/reference.html#sec-ref-slot-config):

``` ini
[keyring]
path=/path/to/keyring.pem
```

**Reply by tigerinus** [reply]
Bear with my limited knowledge, but how do I create a `keyring.pem`? I couldn't find any content in the buildroot doc. Thanks!

**Reply by ejoerns** [reply]
Depends on how you created your keys.  If you have just a self-signed key, then you can use the certificate/pubkey as keyring.

For more complex setups, a keyring can be generated by concatenating accepted certificates.

We have an example for derived keys (with the result being in the git, too) under https://github.com/rauc/rauc/blob/master/test/openssl-ca-create.sh

**Reply by tigerinus** [reply]
Can you elaborate on how to use certificate/pubkey as keyring? Do I just combine two `.pem` files as one?

I tried running the shell script, and found it has generated many `.pem` files. I can tell `*-ca.pem` files are keyring files, but which files are cert and key files?

Thanks.

**Reply by ejoerns** [reply, marked as answer]
> Can you elaborate on how to use certificate/pubkey as keyring? Do I just combine two .pem files as one?

Basically, yes. Or just use a single one. This depends on your setup and hierarchy.

For more details about how to set up keys and how to deal with X.509 setup, I fear I have to point you to general documentation on this topic.

If you do not have any special requirements, yet, maybe also what we have in `meta-rauc` (Yocto) might help to get a picture of what to do: https://github.com/rauc/meta-rauc/tree/master/scripts
